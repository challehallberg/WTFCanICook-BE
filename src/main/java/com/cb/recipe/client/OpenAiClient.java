package com.cb.recipe.client;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@Component
public class OpenAiClient {
    private final WebClient client;

    public OpenAiClient(@Qualifier("openAiWebClient") final WebClient webClient) {
        this.client = webClient;
    }

    public String processImage(final String requestBody) {
        //TODO(CB): Add retryability
        return client.post()
                .uri("/chat/completions")
                .bodyValue(requestBody)
                .retrieve()
                .bodyToMono(String.class)
                .block();
    }
}
