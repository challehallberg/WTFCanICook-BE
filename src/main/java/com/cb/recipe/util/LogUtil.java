package com.cb.recipe.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class LogUtil {
    private static final ObjectMapper objectMapper = new ObjectMapper();
    public static void logForMe(Object toBeLogged) {
        try {
            log.info(objectMapper.writeValueAsString(toBeLogged));
        } catch (JsonProcessingException e) {
            log.error("Error logging object.");
        }
    }
}
