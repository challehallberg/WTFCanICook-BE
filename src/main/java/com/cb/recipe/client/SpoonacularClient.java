package com.cb.recipe.client;

import com.cb.recipe.model.spoonacular.instruction.InstructionInner;
import com.cb.recipe.model.spoonacular.recipe.RecipeInner;
import com.cb.recipe.model.spoonacular.recipeInformation.RecipeInformationInner;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;

@Component
public class SpoonacularClient {

    private final WebClient client;

    //TODO(CB): Fix retryability
    public SpoonacularClient(@Qualifier("spoonacularWebClient") final WebClient webClient) {
        this.client = webClient;
    }


    /**
     * Calls GET /recipes/findByIngredients endpoint with the provided parameters.
     *
     * @param ingredients  a comma-separated list of ingredients (e.g., "apples,flour,sugar")
     * @param number       the maximum number of recipes to return
     * @param ranking      whether to maximize used ingredients (1) or minimize missing ingredients (2)
     * @param ignorePantry whether to ignore typical pantry items (true/false)
     * @return a Mono containing a list of Recipe objects
     */
    public List<RecipeInner> getRecipesByIngredients(final String ingredients, final int number, final int ranking, final boolean ignorePantry) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/findByIngredients")
                        .queryParam("ingredients", ingredients)
                        .queryParam("number", number)
                        .queryParam("ranking", ranking)
                        .queryParam("ignorePantry", ignorePantry)
                        .queryParam("limitLicense", true)
                        .build())
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<RecipeInner>>() {
                })
                .block();
    }

    /**
     * Retrieves bulk recipe information for the given recipe IDs.
     * <p>
     * This method calls the Spoonacular API endpoint "/recipes/informationBulk" using the provided query parameters.
     * It accepts a comma-separated string of recipe IDs and a flag indicating whether to include nutrition data.
     * The JSON response is deserialized into a list of {@link RecipeInformationInner} objects.
     * </p>
     *
     * @param ids              A comma-separated string of recipe IDs (e.g., "715538,716429").
     * @param includeNutrition A boolean flag to include nutrition data in the recipe information.
     * @return A list of {@link RecipeInformationInner} objects corresponding to the provided recipe IDs.
     */
    public List<RecipeInformationInner> getRecipeInformationBulk(final String ids, final boolean includeNutrition) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/informationBulk")
                        .queryParam("ids", ids)
                        .queryParam("includeNutrition", includeNutrition)
                        .build())
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<RecipeInformationInner>>() {
                })
                .block();
    }

    /**
     * Retrieves analyzed instructions for a given recipe.
     * <p>
     * This method calls the Spoonacular API endpoint "/recipes/{id}/analyzedInstructions" with the specified recipe id
     * and step breakdown flag. The JSON response is deserialized into a list of {@link InstructionInner} objects.
     * </p>
     *
     * @param id            the recipe id (e.g., 324694)
     * @param stepBreakdown whether to break down the recipe steps even more
     * @return a list of {@link InstructionInner} objects containing the analyzed instructions for the recipe
     */
    public List<InstructionInner> getAnalyzedInstructions(final int id, final boolean stepBreakdown) {
        return client.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/recipes/{id}/analyzedInstructions")
                        .queryParam("stepBreakdown", stepBreakdown)
                        .build(id))
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<List<InstructionInner>>() {
                })
                .block();
    }
}
