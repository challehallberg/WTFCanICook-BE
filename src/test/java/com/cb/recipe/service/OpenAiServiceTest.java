package com.cb.recipe.service;

import com.cb.recipe.client.OpenAiClient;
import com.cb.recipe.util.UnitTest;
import com.github.benmanes.caffeine.cache.Cache;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

@ExtendWith(MockitoExtension.class)
class OpenAiServiceTest extends UnitTest {
    @Mock
    private OpenAiClient client;

    @Mock
    private Cache<String, String> imageResponseCache;

    @Mock
    private MultipartFile file;

    @InjectMocks
    private OpenAiService openAiService;

    @Test
    void givenValidFileWithNoCacheHit_whenProcessImage_thenExpectedResult() throws IOException {
        Mockito.when(file.getOriginalFilename())
                .thenReturn("testFile.png");
        Mockito.when(file.getBytes())
                .thenReturn("Dummy Bytes".getBytes(StandardCharsets.UTF_8));
        Mockito.when(imageResponseCache.getIfPresent("testFile.png"))
                .thenReturn(null);
        Mockito.when(client.processImage(Mockito.anyString()))
                .thenReturn(mockedImageResponse());

        String result = openAiService.processImage(file);

        snapshot(result);
    }

    @Test
    void givenValidFileInCache_whenProcessImage_thenExpectedResultFromCache() {
        Mockito.when(file.getOriginalFilename())
                .thenReturn("cachedFile.png");
        Mockito.when(imageResponseCache.getIfPresent("cachedFile.png"))
                .thenReturn("apple, banana, tomato");

        String result = openAiService.processImage(file);

        snapshot(result);
    }

    private String mockedImageResponse() {
        return """
                {
                  "choices": [
                    {
                      "message": {
                        "content": "apple, banana, tomato, (it might be Harald's fault)"
                      }
                    }
                  ]
                }
                """;
    }
}