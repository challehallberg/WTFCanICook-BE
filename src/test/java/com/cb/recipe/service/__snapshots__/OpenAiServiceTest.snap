com.cb.recipe.service.OpenAiServiceTest.givenValidFileInCache_whenProcessImage_thenExpectedResultFromCache=[
  "apple, banana, tomato"
]


com.cb.recipe.service.OpenAiServiceTest.givenValidFileWithNoCacheHit_whenProcessImage_thenExpectedResult=[
  "apple, banana, tomato, (it might be Harald's fault)"
]